packages/node/src/core/evm/templates/template-verification.ts

```diff
@@ -0,0 +1,89 @@
+import { ethers } from 'ethers';
+import flatMap from 'lodash/flatMap';
+import * as logger from '../../logger';
+import { ApiCall, ApiCallTemplate, ClientRequest, LogsData, RequestErrorCode, RequestStatus } from '../../../types';
+
+interface ApiCallTemplatesById {
+  [id: string]: ApiCallTemplate;
+}
+
+export const TEMPLATE_VALIDATION_FIELDS = [
+  'providerId',
+  'endpointId',
+  'requesterIndex',
+  'designatedWallet',
+  'fulfillAddress',
+  'fulfillFunctionId',
+  'encodedParameters',
+];
+
+export function getExpectedTemplateId(template: ApiCallTemplate): string {
+  const templateValues = TEMPLATE_VALIDATION_FIELDS.map((f) => template[f]);
+
+  return ethers.utils.keccak256(
+    ethers.utils.defaultAbiCoder.encode(
+      ['bytes32', 'bytes32', 'uint256', 'address', 'address', 'bytes4', 'bytes'],
+      templateValues
+    )
+  );
+}
+
+export function verify(
+  apiCalls: ClientRequest<ApiCall>[],
+  templatesById: ApiCallTemplatesById
+): LogsData<ClientRequest<ApiCall>[]> {
+  const logsWithVerifiedApiCalls: LogsData<ClientRequest<ApiCall>>[] = apiCalls.map((apiCall) => {
+    if (!apiCall.templateId) {
+      return [[], apiCall];
+    }
+
+    // If the request is already errored, blocked or ignored, then we don't want to make further
+    // updates to either the status or error code
+    const ignoredStatuses = [RequestStatus.Blocked, RequestStatus.Errored, RequestStatus.Ignored];
+    if (ignoredStatuses.includes(apiCall.status)) {
```

Comment:
Note to self: this should ignore ALL non-pending statuses

Reply:
Why not check for if the request status is pending?

Reply:
I'm going to update this to this which is consistent with other modules:

```ts
if (apiCall !== RequestStatus.Pending) {
  ...
}
```

I want to be very careful with updating statuses as it could potentially have a critical impact if something goes wrong. All API calls start out as "pending" and then can only have their status (and error code) changed once after that. I might make a helper that updates status & error code that also checks that status is pending first.